ext/openssl/openssl.c

```diff
@@ -5410,48 +5447,704 @@ PHP_FUNCTION(openssl_pkcs7_decrypt)
 
 /* }}} */
 
-/* {{{ proto bool openssl_private_encrypt(string data, string &crypted, mixed key [, int padding])
-   Encrypts data with private key */
-PHP_FUNCTION(openssl_private_encrypt)
+/* {{{ CMS S/MIME functions taken from PKCS#7 functions */
+
+/* {{{ proto bool openssl_cms_verify(string filename, int flags [, string signerscerts [, array cainfo [, string extracerts [, string content [, string pk7]]]], sigfile, encoding])
+   Verifies that the data block is intact, the signer is who they say they are, and returns the CERTs of the signers */
+PHP_FUNCTION(openssl_cms_verify)
 {
-	zval *key, *crypted;
-	EVP_PKEY *pkey;
-	int cryptedlen;
-	zend_string *cryptedbuf = NULL;
-	int successful = 0;
-	zend_resource *keyresource = NULL;
-	char * data;
-	size_t data_len;
-	zend_long padding = RSA_PKCS1_PADDING;
+	X509_STORE * store = NULL;
+	zval * cainfo = NULL;
+	STACK_OF(X509) *signers= NULL;
+	STACK_OF(X509) *others = NULL;
+	CMS_ContentInfo * cms = NULL;
+	BIO * in = NULL, * datain = NULL, * dataout = NULL, * p7bout  = NULL, *sigbio = NULL;
+	zend_long flags = 0;
+	char * filename;
+	size_t filename_len;
+	char * extracerts = NULL;
+	size_t extracerts_len = 0;
+	char * signersfilename = NULL;
+	size_t signersfilename_len = 0;
+	char * datafilename = NULL;
+	size_t datafilename_len = 0;
+	char * p7bfilename = NULL;
+	size_t p7bfilename_len = 0;
+	char * sigfile = NULL;
+	size_t sigfile_len = 0;
+	zend_long encoding = ENCODING_CMS;
 
-	if (zend_parse_parameters(ZEND_NUM_ARGS(), "szz|l", &data, &data_len, &crypted, &key, &padding) == FAILURE) {
+	RETVAL_LONG(-1);
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "pl|p!ap!p!p!p!l", &filename, &filename_len,
+				&flags, &signersfilename, &signersfilename_len, &cainfo,
+				  &extracerts, &extracerts_len, &datafilename, &datafilename_len, &p7bfilename, &p7bfilename_len,&sigfile,&sigfile_len,&encoding) == FAILURE) {
 		RETURN_THROWS();
 	}
+
+	in = BIO_new_file(filename, PHP_OPENSSL_BIO_MODE_R(flags));
+	if (in == NULL) {
+	  php_openssl_store_errors();
+	  goto clean_exit;
+	}
+	if ( sigfile && CMS_DETACHED ) {
+	  sigbio = BIO_new_file(sigfile, PHP_OPENSSL_BIO_MODE_R(flags));
+
+	  switch(encoding) {
+	  	case ENCODING_PEM:
+			cms=PEM_read_bio_CMS(sigbio,NULL,0,NULL);
```

Comment:
CS: missing space before and after `=` (the same also few lines below)

Reply:
got it.